### ISSUE TITLE
http-request capture in named defaults causes segfault

### ISSUE THREAD

**Issue by Tristan971** [haproxy/haproxy#1674]
### Detailed Description of the Problem

When using a statement like `http-request capture hdr(Host) len 32` in a named defaults section, requests cause the worker process to exit with code 139 (due to a segfault).

### Expected Behavior

The `http-request` class of statements should work in named frontends as per [the documentation](https://docs.haproxy.org/2.5/configuration.html#4.1).

### Steps to Reproduce the Behavior

1. Make a named defaults section with `http-request capture hdr(Host) len 32`
2. Have a frontend using said named defaults
3. Make a request that will hit the frontend

### Do you have an idea how to solve the issue?

Moving the statement into the frontend fixes the issue

### What is your configuration?

After shrinking it as much as possible, this is the reproducer I came up with:
```haproxy
global
    daemon
    user haproxy
    group haproxy
    log /dev/log local0

defaults defaults-base
    log global
    http-request capture hdr(Host) len 32

frontend http from defaults-base
    bind *:8080
    mode http
    http-request return status 200
```

Then, using the official docker image (note that I observed it using [this PPA](https://launchpad.net/~vbernat/+archive/ubuntu/haproxy-2.5) originally, so it's not specific to Docker or that image) and this sample dockerfile:

```dockerfile
FROM haproxy:2.5
COPY haproxy.cfg /usr/local/etc/haproxy/haproxy.cfg
```

Then build and run it:
```bash
docker build . -t spooky-defaults
docker run -it --name haproxy --rm -p8080:8080 spooky-defaults
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.5.5-384c5c5 2022/03/14 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.5.html
Running on: Linux 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=32).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Last Outputs and Backtraces

Log output
```plain

<134>Apr 25 01:31:44 haproxy[9]: Connect from 172.17.0.1:36620 to 172.17.0.2:8080 (http/HTTP)
[NOTICE]   (1) : haproxy version is 2.5.5-384c5c5
[NOTICE]   (1) : path to executable is /usr/local/sbin/haproxy
[ALERT]    (1) : Current worker (9) exited with code 139 (Segmentation fault)
[ALERT]    (1) : exit-on-failure: killing every processes with SIGTERM
[WARNING]  (1) : All workers exited. Exiting... (139)
```

Backtrace (that one is not from the Docker image but the host I saw it on first, with the exact config from above, as I can't get coredumps to properly work within Docker for some reason)
```plain
$ gdb /usr/sbin/haproxy ./core._usr_sbin_haproxy.114.31d2f1dc-6069-422c-b135-9f3230c862a7.27747.7396470
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04.1) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/sbin/haproxy...
(No debugging symbols found in /usr/sbin/haproxy)
[New LWP 27747]
[New LWP 27749]
[New LWP 27751]
[New LWP 27750]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/h'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000060b80eea6b9e in ?? ()
[Current thread is 1 (Thread 0x73a0d8635f80 (LWP 27747))]
(gdb) t a a bt full

Thread 4 (Thread 0x73a0d6cf7700 (LWP 27750)):
#0  0x000073a0d8a3549e in epoll_wait (epfd=12, events=0x73a0c4025e00, maxevents=200, timeout=60000) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
        resultvar = 18446744073709551612
        sc_cancel_oldtype = 0
        sc_ret = <optimized out>
#1  0x000060b80ed4e633 in ?? ()
No symbol table info available.
#2  0x000060b80eea1c84 in run_poll_loop ()
No symbol table info available.
#3  0x000060b80eea2129 in ?? ()
No symbol table info available.
#4  0x000073a0d8ff5609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {127134635882240, 8171257159924893350, 140729750570398, 140729750570399, 140729750570400, 127134635837312, -7626970739694844250, -7626965924522572122}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#5  0x000073a0d8a35163 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 3 (Thread 0x73a0d62f5700 (LWP 27751)):
#0  0x000073a0d8a3549e in epoll_wait (epfd=17, events=0x73a0c8025e00, maxevents=200, timeout=60000) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
        resultvar = 18446744073709551612
        sc_cancel_oldtype = 0
        sc_ret = <optimized out>
#1  0x000060b80ed4e633 in ?? ()
No symbol table info available.
#2  0x000060b80eea1c84 in run_poll_loop ()
No symbol table info available.
#3  0x000060b80eea2129 in ?? ()
No symbol table info available.
#4  0x000073a0d8ff5609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {127134625388288, 8171257159924893350, 140729750570398, 140729750570399, 140729750570400, 127134625343360, -7626969916134865242, -7626965924522572122}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#5  0x000073a0d8a35163 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 2 (Thread 0x73a0d862a700 (LWP 27749)):
#0  0x000073a0d8a3549e in epoll_wait (epfd=9, events=0x73a0d00264f0, maxevents=200, timeout=60000) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
        resultvar = 18446744073709551612
        sc_cancel_oldtype = 0
        sc_ret = <optimized out>
#1  0x000060b80ed4e633 in ?? ()
No symbol table info available.
#2  0x000060b80eea1c84 in run_poll_loop ()
No symbol table info available.
#3  0x000060b80eea2129 in ?? ()
No symbol table info available.
#4  0x000073a0d8ff5609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {127134662305536, 8171257159924893350, 140729750570398, 140729750570399, 140729750570400, 127134662260608, -7626964871158905178, -7626965924522572122}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#5  0x000073a0d8a35163 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.

Thread 1 (Thread 0x73a0d8635f80 (LWP 27747)):
#0  0x000060b80eea6b9e in ?? ()
No symbol table info available.
#1  0x000060b80edca049 in ?? ()
No symbol table info available.
#2  0x000060b80edccdeb in http_process_req_common ()
No symbol table info available.
#3  0x000060b80edfe5e0 in process_stream ()
--Type <RET> for more, q to quit, c to continue without paging--
No symbol table info available.
#4  0x000060b80eed199b in run_tasks_from_lists ()
No symbol table info available.
#5  0x000060b80eed2103 in process_runnable_tasks ()
No symbol table info available.
#6  0x000060b80eea1cc4 in run_poll_loop ()
No symbol table info available.
#7  0x000060b80eea2129 in ?? ()
No symbol table info available.
#8  0x000060b80ed4aba2 in main ()
No symbol table info available.
```

### Additional Information

The other similar statements I use in named defaults do work just fine, that is http-request actions:
- set-var
- add/del/set-header
- deny
- track-sc0

Same for http-after-response add/set-header

So it *seems* limited to capture, but I have not tried every single possible action, nor am I familiar with HAProxy's internals enough to know which ones would be pertinent to test.

**Comment by capflam** [comment]
I can confirm the issue. I'll try to fix it for the next release. Thanks !

**Comment by capflam** [comment]
I pushed a fix. It should be ok now. I'll backport it to 2.5 very soon. A new release is planned this week

**Comment by Tristan971** [comment]
That was quick, thanks!

**Comment by wtarreau** [comment]
Both fixes are now in 2.5.6, let's close.
